### PR TITLE
[Aio] Generate gRPC AsyncIO API Reference

### LINKED ISSUE #21400: [AIO] API reference document generation
Generates the API reference for Async API. It should include:
* All newly added classes / functions
* Re-used existing APIs (e.g. security stuffs)
* Documentation in `__init__.py` which states the caveats

### PR DESCRIPTION
**Pull request by lidizheng**
Fixes #21400

This PR enhances the document generator to generate API reference. It also fills-in many missing interfaces which we didn't specify during development.

Now, `__init__.py` only contains import statements, and an `__all__` statement that exposes limited API. Interfaces are created in separate files (majority copied from our gRFC):
* `_base_call.py`: Call object related interfaces
* `_base_channel.py`: Channel related interfaces
* `_base_server.py`: server-side interfaces

Preview of API reference website:
~~https://storage.googleapis.com/lidiz_static_api_doc/grpc_api_reference.tar.gz~~
https://storage.googleapis.com/lidiz_static_api_doc/grpc_api_reference_v2.tar.gz

### CODE REVIEW

File: doc/python/sphinx/grpc_asyncio.rst [hunk 1]
```diff
@@ -0,0 +1,120 @@
+gRPC AsyncIO API
+================
+
+.. module:: grpc.experimental.aio
+
+Overview
+--------
+
+gRPC AsyncIO API is the **new version** of gRPC Python whose architecture is
+tailored to AsyncIO. Underlying, it is using C-Core's callback API, and
```
Comment (gnossen) [inline doc/python/sphinx/grpc_asyncio.rst]: This should be written for an audience who is not familiar with the internals of gRPC. If we use jargon ("Core", "callback API"), we either need to explain what it means in-line or link to something that *does* explain what it means.

File: doc/python/sphinx/grpc_asyncio.rst [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+-------
+
+gRPC Async API objects may only be used on the thread on which they were
+created. AsyncIO doesn't provide thread safety for most of its APIs.
+
+
+Module Contents
+---------------
+
+Turn-On AsyncIO Mode
```
Comment (gnossen) [inline doc/python/sphinx/grpc_asyncio.rst]: Suggestion: "Enable AsyncIO in gRPC". Here and throughout.

File: doc/python/sphinx/grpc_asyncio.rst [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+^^^^^^^^^^^^^^^^^^^^
+
+.. function:: init_grpc_aio
+
+    Turn-on AsyncIO mode for gRPC Python.
+
+    This function is idempotent, and it should be invoked before creation of
```
Comment (gnossen) [inline doc/python/sphinx/grpc_asyncio.rst]: Comma splice. No comma necessary.

File: doc/python/sphinx/grpc_asyncio.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    AsyncIO stack objects. Otherwise, the application might deadlock.
+
+    This function enables AsyncIO IO manager and disables threading for entire
```
Comment (gnossen) [inline doc/python/sphinx/grpc_asyncio.rst]: "IO manager" is gRPC-specific jaron. Please either explain in line or permalink to something that does. Also, we should use a definite article here.

File: doc/python/sphinx/grpc_asyncio.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    process. After this point, there should not be blocking calls unless it is
```
Comment (gnossen) [inline doc/python/sphinx/grpc_asyncio.rst]: s/for entire process/for the entire process/

File: doc/python/sphinx/grpc_asyncio.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    taken cared by AsyncIO.
```
Comment (gnossen) [inline doc/python/sphinx/grpc_asyncio.rst]: Suggestion: "After invoking this function, making blocking function calls will block the event loop, potentially starving all RPCs in the process. Refer to the Python language documentation on AsyncIO for more details (https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)".

File: doc/python/sphinx/grpc_asyncio.rst (hunk 4 above)
Comment (pfreixes) [inline doc/python/sphinx/grpc_asyncio.rst]: disables threading? the user might get it in the wrong way, the user is free to create new threads after the initialization of the Asyncio IO manager.

You meant that gRPC library would be executed using only one thread? if so I would rephrase a bit the content for making it more clear.

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+import grpc
+
+from . import _base_call
+from ._typing import DeserializingFunction, MetadataType, SerializingFunction
+
+_IMMUTABLE_EMPTY_TUPLE = tuple()
+
+
+class UnaryUnaryMultiCallable(abc.ABC):
+    """Factory an asynchronous unary-unary RPC stub call from client-side."""
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Suggestion: "Factory for an asynchronous unary RPC stub call on the client-side."

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+        Raises:
+          RpcError: Indicating that the RPC terminated with non-OK status. The
+            raised RpcError will also be a Call for the RPC affording the RPC's
+            metadata, status code, and details.
+        """
+
+
+class UnaryStreamMultiCallable(abc.ABC):
+    """Affords invoking a unary-stream RPC from client-side in an asynchronous way."""
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: I know "Affords invoking" has been there forever in the existing stack, but it's always been quite unnatural to my ear. How about "Enables asynchronous invocation of a server-streaming RPC ."? I think "client-side" is redundant. Invocation only happens on the client side.

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+          compression: An element of grpc.compression, e.g.
+            grpc.compression.Gzip. This is an EXPERIMENTAL option.
+
+        Returns:
+          A Call object instance which is an awaitable object.
+        """
+
+
+class StreamUnaryMultiCallable(abc.ABC):
+    """Affords invoking a stream-unary RPC from client-side in an asynchronous way."""
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Suggestion: "Enables asynchronous invocation of a client-streaming RPC."

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+
+        Raises:
+          RpcError: Indicating that the RPC terminated with non-OK status. The
+            raised RpcError will also be a Call for the RPC affording the RPC's
+            metadata, status code, and details.
+        """
+
+
+class StreamStreamMultiCallable(abc.ABC):
+    """Affords invoking a stream-stream RPC from client-side in an asynchronous way."""
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Suggestion: "Enables asynchronous invocation of a bidirectional streaming RPC."

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+            service-side of the RPC.
+          credentials: An optional CallCredentials for the RPC. Only valid for
+            secure Channel.
+          wait_for_ready: This is an EXPERIMENTAL argument. An optional
+            flag to enable wait for ready mechanism
+          compression: An element of grpc.compression, e.g.
+            grpc.compression.Gzip. This is an EXPERIMENTAL option.
+
+        Returns:
+          A Call object instance which is an awaitable object.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Raises?

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+
+        Raises:
+          RpcError: Indicating that the RPC terminated with non-OK status. The
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: s/Indicating/Indicates/

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: s/Indicating/Indicates/

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: s/Indicating/Indicates/

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+            raised RpcError will also be a Call for the RPC affording the RPC's
+            metadata, status code, and details.
+        """
+
+
+class Channel(abc.ABC):
+    """Asynchronous Channel implementation.
+
+    A cygrpc.AioChannel-backed implementation.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Jargon. In this case, we shouldn't mention Cython at all. It's purely an implementation detail. How about just "An asynchronous channel"?

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+        been terminated are cancelled. If a grace period is not specified
+        (by passing None for grace), all existing RPCs are cancelled immediately.
+
+        This method is idempotent.
+        """
+
+    @abc.abstractmethod
+    def get_state(self,
+                  try_to_connect: bool = False) -> grpc.ChannelConnectivity:
+        """Check the connectivity state of a channel.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: s/Check/Checks/

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+
+        Returns: A ChannelConnectivity object.
+        """
+
+    @abc.abstractmethod
+    async def wait_for_state_change(
+            self,
+            last_observed_state: grpc.ChannelConnectivity,
+    ) -> None:
+        """Wait for a change in connectivity state.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: s/Wait/Waits/

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+
+        This is an EXPERIMENTAL API.
+
+        The function blocks until there is a change in the channel connectivity
+        state from the "last_observed_state". If the state is already
+        different, this function will return immediately.
+
+        There is an inherent race between the invocation of
+        "Channel.wait_for_state_change" and "Channel.get_state". The state can
+        change arbitrary times during the race, so there is no way to observe
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_channel.py]: s/arbitrary times/arbitrarily many times/

File: src/python/grpcio/grpc/experimental/aio/_base_server.py [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+        This method is only safe to call before the server is started.
+
+        Args:
+          generic_rpc_handlers: A sequence of GenericRpcHandlers that will be
+          used to service RPCs.
+        """
+
+    @abc.abstractmethod
+    def add_insecure_port(self, address: str) -> int:
+        """Opens an insecure port for accepting RPCs.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_server.py]: We probably want to add a brief explanation of what we man by "port". The term is a bit overloaded.

File: src/python/grpcio/grpc/experimental/aio/_base_server.py [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+        (i.e. the most restrictive grace value is used).
+
+        Args:
+          grace: A duration of time in seconds or None.
+        """
+
+    @abc.abstractmethod
+    async def wait_for_termination(self,
+                                   timeout: Optional[float] = None) -> bool:
+        """Block current coroutine until the server stops.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_server.py]: This doesn't actually *block*, right? Someone might see the word "block" and get worried about using this on their event loop.

File: src/python/grpcio/grpc/experimental/aio/_base_server.py [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+
+
+class ServicerContext(Generic[RequestType, ResponseType], abc.ABC):
+    """A context object passed to method implementations."""
+
+    @abc.abstractmethod
+    async def read(self) -> RequestType:
+        """Reads one message from the RPC.
+
+        Only one read operation is allowed simultaneously. Mixing new streaming API and old
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_server.py]: What is the "new streaming API"? What is the "old streaming API"? The audience will have much less context than you and I do.

File: src/python/grpcio/grpc/experimental/aio/_base_server.py [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+        streaming API will resulted in undefined behavior.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_server.py]: s/resulted/result/

File: src/python/grpcio/grpc/experimental/aio/_base_server.py [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+        This method need not be called by method implementations if they wish
+        the gRPC runtime to determine the status code of the RPC.
+
+        Args:
+          code: A StatusCode object to be sent to the client.
+        """
+
+    @abc.abstractmethod
+    def set_details(self, details: str) -> None:
+        """Sets the value to be used as detail string upon RPC completion.
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_server.py]: s/as detail string/as the detail string/

File: doc/python/sphinx/grpc_asyncio.rst [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+
+
+Create Client
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply doc/python/sphinx/grpc_asyncio.rst]: Updated to `Channel`. However, I found it not easy to explain the relation between Channel objects and gRPC client without mentioning many other concepts. Here is my wordy attempt:

```
Channels are the abstraction of clients, where most of networking logic
happens, for example, managing one or more underlying connections, name
resolution, load balancing, flow control, etc.. If you are using ProtoBuf,
Channel objects works best when further encapsulate into stub objects, then the
application can invoke remote functions as if they are local functions.
```

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 7 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Updated to `unary call`/`client streaming`/`server streaming`/`bidirectional streaming` terminology.

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 12 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Good catch.

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 8 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Changed all docstring for MultiCallable objects, and removed "client-side".

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 11 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Updated. Good catch!

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 9 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Updated.

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 15 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Good catch, I missed this docstring. I removed Cython specific description, and adding details about `async with`:

```Python
class Channel(abc.ABC):
    """Enables asynchronous RPC invocation as a client.

    Channel objects implement the Asynchronous Context Manager (aka. async
    with) type, although they are not supportted to be entered and exited
    multiple times.
    """
```

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 16 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Ah, the C++ readability.

File: src/python/grpcio/grpc/experimental/aio/_base_channel.py (hunk 18 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_channel.py]: Updated, but why there is a difference here? I thought arbitrary times means number of occurrence is equal or larger than zero.

File: src/python/grpcio/grpc/experimental/aio/_base_server.py (hunk 19 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_server.py]: Inserts a brief description of the term `port`, also included UDP. Someday, we might support UDP, since Core supports it.

```
        A port is a communication endpoint that used by networking protocols,
        like TCP and UDP. To date, we only support TCP.
```

File: src/python/grpcio/grpc/experimental/aio/_base_server.py (hunk 20 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_server.py]: Good catch. I flipped the description to use `continues`:

```
Continues current coroutine once the server stops.
```

File: src/python/grpcio/grpc/experimental/aio/_base_server.py (hunk 21 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_server.py]: The restriction removed. Since for server-side, there isn't a clean way to remove the supplied `request_async_iterator` in method signature. It can be ironic to prohibit the mix of usage while explicitly giving the option.

Once we had a better solution, we can revisit this docstring. E.g., we could support two different method handler signature.

File: src/python/grpcio/grpc/experimental/aio/_base_server.py (hunk 23 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_server.py]: Good catch!

**Comment by lidizheng** [comment]
@gnossen PTALA.
